lnrpc/routerrpc/router.proto

```diff
@@ -818,12 +818,18 @@ message ForwardHtlcInterceptResponse {
     // The preimage in case the resolve action is Settle.
     bytes preimage = 3;
 
-    // Encrypted failure message in case the resolve action is Fail.
+    // Failure message in case the resolve action is Fail. The field
+    // failure_message_unencrypted indicates whether this message is already
+    // encrypted for the first hop.
     //
-    // If failure_message is specified, the failure_code field must be set
-    // to zero.
+    // If failure_message is specified, the failure_code field must be set to
+    // zero.
     bytes failure_message = 4;
 
+    // Indicates whether the failure message still needs to be encrypted for the
+    // first hop.
+    bool failure_message_unencrypted = 6;
```

Comment:
>yes exactly, what i mean is to do this at the API level in an unencrypted way, to avoid the caller having to build the encrypted failure message

Okay got it. I think it is debatable whether this is functionality that should live inside lnd. Passing in the private key is a security risk too.

Reply:
This discussion thread still relevant?

Reply:
It is a good point, and indeed just passing in the shared secret for the virtual hop and letting lnd do the encryption could be a middle way.

This PR is about adding support for unencrypted failure reasons though, not making it easier to use the encrypted reason functionality that is pre-existing. I think that would be more suitable as a follow-up? In that a `failure_message_shared_secret` could be added to indicate that `failure_message` still has to be encrypted. At that point, `failure_message_unencrypted = false` can be deprecated.

Reply:
@joostjager IIUC, this is intended for a use case for "virtual nodes", so anycast-like behavior using special hop hints. In that case, how does this help? It'll let you send a _plaintext_ error (from this interceptor sub-system) and have lnd encrypted that, where it's the source node. IIUC though, you actually want to be able to send errors from the "virtual node", which means you need to use the shared secret for _that_ hop for the path finding retry/apply logic to work properly.

Reply:
@joostjager IIUC, this is intended for a use case for "virtual nodes", so anycast-like behavior using special hop hints. In that case, how does this help? It'll let you send a _plaintext_ error (from this interceptor sub-system) and have lnd encrypted that, where it's the source node. IIUC though, you actually want to be able to send errors from the "virtual node", which means you need to use the shared secret for _that_ hop for the path finding retry/apply logic to work properly.

Reply:
>IIUC though, you actually want to be able to send errors from the "virtual node", which means you need to use the shared secret for that hop for the path finding retry/apply logic to work properly.

That is correct.

But sometimes though the virtual node may want to return an error originating from the preceding real lnd node. An example of this is when there is a routing fee associated with the virtual channel. https://github.com/NYDIG-OSS/lnmux/pull/101 illustrates this. The real node has to encrypt the insufficient fee failure message.

Intercepted htlcs do not pass through the switch in the real lnd instance. This means that any forwarding logic (such as the fee check) needs to be replicated by the external interceptor, and the interceptor requires the ability to instruct the real node to produce certain failure messages. 

This has been possible in master for some time (https://github.com/lightningnetwork/lnd/blob/bf5aab9d5246daede3f499af880a5d7be6c6ee88/lnrpc/routerrpc/router.proto#L930) using the `failure_code` field. This however is not a complete API. Ideally the complete failure message structure would need to be replicated on the proto level to be able to signal any failure. This PR takes the approach of accepting an encoded but not yet encrypted failure message to avoid that.

In summary this PR:
* Completes the interceptor API by allowing users to supply a full failure message rather than a restrictive set of failure codes without additional data fields.
* Enable use cases that require different failure messages to be returned, such as a non-zero virtual channel fee.